### ISSUE TITLE
600 line Rust source file takes more than 2 minutes

### ISSUE THREAD

**Issue by robinst** [trishume/syntect#33]
How to reproduce ([link to view file](https://github.com/rust-lang/rust/blob/a6fe6c9be10c79d64c988d490519f5cc07e9060d/src/librustc_borrowck/borrowck/mir/dataflow/impls.rs)):

```
curl -LO 'https://github.com/rust-lang/rust/raw/a6fe6c9be10c79d64c988d490519f5cc07e9060d/src/librustc_borrowck/borrowck/mir/dataflow/impls.rs'
time target/release/examples/synstats impls.rs
```

Running with syntect from master (currently 431174d97fe8df2fc798e11c96991e989b23a1e9), this is the output of `time`:

```
target/release/examples/synstats impls.rs  141.49s user 0.54s system 99% cpu 2:22.71 total
```

Sublime text build 3125 seems to have no problem with the file and loads instantly.

**Comment by trishume** [comment]
Whoah that's messed up. And even with release mode. I wonder if that's a regression or if all versions do badly on that file.

I'll try and take a look at this soon. Unfortunately I have school assignments, but I may find time this week, as well as for finishing #27.

**Comment by robinst** [comment]
Thanks. Tried with 1.0.0 (7981ad3dfabb0d7921466b064f9e79f6f9bd4f50) just now, same problem (153.66s).

**Comment by jmacdonald** [comment]
I think I've run into this; the Rust highlighter performance is terrible with one particular file (I've not run into this before, and have used syntect pretty extensively). [Here](https://gist.github.com/jmacdonald/78aa5b254c9824a96a9177de4beb7b4c)'s the file I was highlighting.

I profiled my application to isolate the poorly performing function, and it appears to be in the `onig` library:
![screen shot 2017-01-31 at 10 39 44 pm](https://cloud.githubusercontent.com/assets/458253/22494554/57af9f1e-e806-11e6-9863-17c63e72c592.png)

Happy to provide other details and/or help reproduce, as necessary!

**Comment by trishume** [comment]
@jmacdonald Interesting. From what I remember of previous profiling, 50% of the time being taken in the Regex engine is pretty typical, it's in fact lame that it isn't more of the total run time, but the reengineering to fix that would be tricky. 

In fact if anything the fact that it is only 50% tells me that it is probably not just a super bad regex that takes exponential time or something like that. It's probably some set of parser states that is doing way too much work per character of the file parsed, maybe going over it a bunch of times when it otherwise shouldn't.

**Comment by jmacdonald** [comment]
@trishume Ah, that profiling was done _manually_, driving a text editor application, so take those percentages with a grain of salt (a large part of the total application runtime was relatively idle while I drove the UI). The `15s` in `match_at` is actually quite poor; it's a result of iterating over the tokens of the specified document ~6 times.

**Comment by robinst** [comment]
Found another problematic file here: https://gist.github.com/robinst/b48dd2f66312ece6a6204e0135fcc60a

It takes 150.41s. I can probably work around that one by not feeding syntect any lines that are longer than, say 500 characters.

**Comment by trishume** [comment]
@robinst oh yah that's a known problem with Sublime's highlighting model. I think Sublime has a long line threshold as well. Notice Github also doesn't bother trying to highlight it.

It's things that don't have absurdly long lines that work fine in Sublime that I definitely consider a bug.

**Comment by jmacdonald** [comment]
@trishume I noticed something peculiar about how the lexer performance degrades, [specifically with the file I previously posted](https://gist.github.com/jmacdonald/78aa5b254c9824a96a9177de4beb7b4c). The `impl ApplicationPreferences` line seems to be where things go wrong. Essentially, the more text I add _after_ the `impl` token, the worse the performance gets, and it seems to degrade exponentially. Maybe there's something in the Rust grammar, specifically around `impl` tokens, that triggers the issue?

**Comment by trishume** [comment]
@jmacdonald thanks that's definitely useful information. I'm going to see if I can take a look at this bug this evening or this weekend.

**Comment by trishume** [comment]
Okay so I did some investigating and it seems Rust is spending all of its time in oniguruma on a single regex search for a fancy regex that must be triggering the worst case time of backtracking regex engines:

`(?=\s*(?:::|(?:[[:alpha:]][_[:alnum:]]*|_[_[:alnum:]]+)|\$|<)+(<.*?>)?\s+for\s+)`

I'm not sure if there's much I can do to fix this without changing the Rust syntax definition or implementing #12. I'm going to look more at the regex to see if I can figure out why it is so bad and what triggers it.

**Comment by trishume** [comment]
The issue can be reproduced with just this:
```rust
// issue #33: it prints this line and then hangs on the next one, hang time increases super-linearly with line length
impl ApplicationPreferenceseeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee {
}
```

The line in `Rust.sublime-syntax` that triggers it is https://github.com/sublimehq/Packages/blob/master/Rust/Rust.sublime-syntax#L627. The first pattern of the `impl-for` context.

**Comment by trishume** [comment]
So I have two possible fixes:

### The patch
Make a PR to the sublimehq/Packages repo to change the regex to 

`(?=\s*(?:::|[[:alpha:]][_[:alnum:]]*+|_[_[:alnum:]]++|\$|<)++(<.*?>)?\s+for\s+)` 

which uses [possessive quantifiers](http://www.regular-expressions.info/possessive.html) to avoid catastrophic backtracking.

This is way easier, but it only solves this specific case, and it is possible they would be reluctant to accept it because it is uglier and it may actually cause real Sublime to use oniguruma instead of their fancy regex engine. The same fact that Sublime's fancy regex engine may not support possessives also means the `regex` crate doesn't, which means that this isn't good if syntect implements the second option:

### The fix

Something I've been meaning to do anyway is to use the `regex` crate for any regex that works with it. Oniguruma would still be needed for things that use fancy features, but many regexes could be run faster. This particular regex uses lookahead, which is a feature `regex` doesn't support, so we would either need to use @raphlinus's `fancy-regex` or add a special case for when the entire regex is in a lookahead (which is common and easy to handle).

This might be a lot of work and I'm not sure when and if I'll have the time/motivation to do it. But it would probably be a better solution and improve performance across the board.

**Comment by jmacdonald** [comment]
I'd love to see syntect move towards `regex` crate adoption. From what I've read, [fancy regex already defers to regex whenever possible](https://www.reddit.com/r/rust/comments/4e0bxx/new_fancyregex_crate_in_development/), so if that crate is in a usable state, it might be best to work on integrating that, and let it handle the underlying library delegation.

I know that's a bigger commitment, and that if you take it on, it won't be an overnight fix. In the interim, another option would be to simply patch the Rust syntax definition dump included with syntect as a temporary workaround.

**Comment by robinst** [comment]
So just to clear up my understanding of the issue: Sublime doesn't hit this issue because it uses the sregex library for this particular regex, which doesn't have the problem?

**Comment by robinst** [comment]
(Also, anyone looking for the source of fancy-regex, it was just split out into its own repository here: https://github.com/google/fancy-regex)

**Comment by trishume** [comment]
@robinst exactly.

So I just talked to @raphlinus over email and he went and [added atomic groups and possessives](https://github.com/google/fancy-regex/commit/8e0c1f8422235c9b9710f2a4722b53ad8bc0d6f8)! 😀 

This is awesome because it's the last real thing that was keeping syntect on Oniguruma. I'm going to try to port syntect to `fancy-regex` on Wednesday, probably in a branch. Then I'll test it:

- If it works incredibly well I'll just merge it and ditch Oniguruma for good. This is unlikely given how little testing fancy-regex has had, there will probably be bugs.
- If it works for some syntaxes but not others, I'll put it behind a Cargo feature so that consumers of the library can decide. Then I'll try and fix bugs in fancy-regex.
- If it doesn't really work for any syntaxes I'll leave it in a branch and try and fix the bugs in fancy-regex.

**Comment by jmacdonald** [comment]
Hey, this is fantastic! 🎉 

I use syntect for syntax highlighting in an editor on a daily basis, so I'm happy to test drive this implementation and file any bugs I encounter. Let me know when it's in a usable state and I'll give it a go!

**Comment by trishume** [comment]
This particular issue was solved by #56 by changing the regex to not exhibit catastrophic backtracking, but catastrophic backtracking is still an issue in general. I'm going to close this issue in favour of moving discussion to #34.
